python/pyspark/sql/types.py

```diff
@@ -938,12 +1016,17 @@ def _infer_type(obj):
                 return MapType(_infer_type(key), _infer_type(value), True)
         else:
             return MapType(NullType(), NullType(), True)
-    elif isinstance(obj, (list, array)):
+    elif isinstance(obj, list):
         for v in obj:
             if v is not None:
                 return ArrayType(_infer_type(obj[0]), True)
         else:
             return ArrayType(NullType(), True)
+    elif isinstance(obj, array):
+        if obj.typecode in _array_type_mappings:
+            return ArrayType(_array_type_mappings[obj.typecode](), False)
+        else:
+            raise TypeError("not supported type: array(%s)" % obj.typecode)
```

Comment:
If we fall back to `_infer_type`, then there should be some dirty changes in test cases to make it pass: Consider the following question:
1. Should we add 'L' as exception for python2 in unsupported types tests, or do we just completely remove unsupported tests?
2. Should we test 'L' for python 2?

I really like how the tests now are organized and these changes above will makes the test very messy. My opinion is, we are not changing the status of 'L' from "supported" to "unsupported", but from "undefined support status" to "unsupported". If changing from "undefined support status" to "unsupported" sounds bad, instead of making these changes to the test cases, I would rather to solve this problem now and keep the test cases clean.